### PR TITLE
Use `dashmap::DashSet` in `uv_extract::unzip`

### PR DESCRIPTION
**Pull request by DaniPopes**
## Summary

Use `dashmap::HashSet` over `Mutex<HashSet>`.
Suggested in https://github.com/astral-sh/uv/pull/3435#discussion_r1592917098.

This is a bit more complex that I thought: the original hashmap lock was held through the `create_dir_all` call, and if we naively swap it for dashmap this is not the case anymore which creates a race condition. See f9a5feaa5b39892364c3f8b63f9ba22aa7917a75.

I don't know if this is worth it, feel free to close if not.

### CODE REVIEW

File: crates/uv-extract/src/stream.rs [hunk 1]
```diff
@@ -13,11 +13,7 @@ use crate::Error;
 /// This is useful for unzipping files as they're being downloaded. If the archive
 /// is already fully on disk, consider using `unzip_archive`, which can use multiple
 /// threads to work faster in that case.
-pub async fn unzip<R: tokio::io::AsyncRead + Unpin>(
-    reader: R,
-    target: impl AsRef<Path>,
```
Comment (BurntSushi) [inline crates/uv-extract/src/stream.rs:18]: Why this change?

File: crates/uv-extract/src/sync.rs [hunk 2]
```diff
@@ -1,21 +1,37 @@
...
+use dashmap::mapref::entry::Entry;
 use rayon::prelude::*;
-use rustc_hash::FxHashSet;
+use rustc_hash::FxHasher;
 use zip::ZipArchive;
 
 use crate::vendor::{CloneableSeekableReader, HasLength};
 use crate::Error;
 
+type FxDashMap<K, V> = dashmap::DashMap<K, V, BuildHasherDefault<FxHasher>>;
```
Comment (BurntSushi) [inline crates/uv-extract/src/sync.rs:12]: Since we're using this as a set, I'd find it clearer to define this as a `FxDashSet<V>` and fix the value type to `()`.

Comment (BurntSushi) [reply crates/uv-extract/src/sync.rs:12]: I think I'd also add a comment here saying why we aren't using a `DashSet` directly. I assume it's because a `DashSet` has no API for holding the lock in the case of a non-existent key.

**Review by BurntSushi — COMMENTED** [review]
I think a `Mutex<HashSet>` is simpler here, so I think I want to see a reason to use a `DashMap` here in lieu of the simpler thing. And in particular, since we're holding the lock across I/O boundaries anyway, it's not clear to me that a concurrent hashmap is going to help us much if at all.

@ibraheemdev Did you have any thoughts here?

**Comment by ibraheemdev** [comment]
Hmm I agree that if we're holding the lock during the fs operation `DashMap` probably doesn't buy us much. We could use a `OnceMap` here to avoid contention on the lock, however, I'm not sure if the contention is a bottleneck given that the download still happens in parallel. If it's not I agree this is probably not worth it.

**Comment by charliermarsh** [comment]
Alright, let's close for now then. Sorry for the churn @DaniPopes, appreciate your work!
